File: 2020/day09/puzzle9.ts
const input: string = await Deno.readTextFile("puzzle9.in");
const preambleLength = 25;

const numbers = input.split("\n").map((line: string) => parseInt(line));

const hasMatchingNumber = (numbers: number[], target: number) => (numberA: number) =>
  numbers.some((numberB: number) => numberA + numberB === target);

function isNumberInvalid(x: number, idx: number) {
  const numbersToCheck = numbers.slice(idx, idx + preambleLength);
  return !numbersToCheck.some(hasMatchingNumber(numbersToCheck, x));
}

function part1() {
  return numbers.slice(preambleLength).find(isNumberInvalid);
}

function part2() {
  const invalidNumberIndex = numbers.slice(preambleLength).findIndex(isNumberInvalid);
  const invalidNumber = numbers[invalidNumberIndex + preambleLength];

  for (let i = 0; i < invalidNumberIndex; i++) {
    let sum = 0;
    let smallest = numbers[i];
    let largest = numbers[i];

    for (let j = i + 1; j < invalidNumberIndex && sum < invalidNumber; j++) {
      sum += numbers[j];

      if (numbers[j] > largest) {
        largest = numbers[j];
      }
      if (numbers[j] < smallest) {
        smallest = numbers[j];
      }

      if (sum == invalidNumber) {
        return smallest + largest;
      }
    }
  }
}

console.log(part1());
console.log(part2());

File: 2020/day09/puzzle9.py
file = open("puzzle9.in")

preamble_length = 25
numbers = []

for line in file:
  number = int(line.strip())
  numbers.append(number)

def is_invalid_number(number_index):
  for i in range(number_index - preamble_length, number_index - 1):
    for j in range(i + 1, number_index):
      if numbers[i] + numbers[j] == numbers[number_index]:
        return True
  return False

def find_invalid_number():
  for i in range(preamble_length, len(numbers)):
    if not is_invalid_number(i):
      return i

def part1():
  invalid_number_index = find_invalid_number()
  return numbers[invalid_number_index]

def part2():
  invalid_number_index = find_invalid_number()
  invalid_number = numbers[invalid_number_index]

  for i in range(0, invalid_number_index - 2):
    s = 0
    smallest = numbers[i]
    largest = numbers[i]
    for j in range(i, invalid_number_index):
      s += numbers[j]

      if numbers[j] > largest:
        largest = numbers[j]
      if numbers[j] < smallest:
        smallest = numbers[j]

      if s == invalid_number:
        return smallest + largest
      elif s > invalid_number:
        break

print(part1())
print(part2())
